storage/src/mmr/operation.rs

```diff
@@ -0,0 +1,267 @@
+        }
+
+        Type::Update(V::try_from(&self.data[K::LEN_ENCODED + 1..]).unwrap())
+    }
+
+    pub fn to_value(&self) -> Option<V> {
+        match self.to_type() {
+            Type::Deleted => None,
+            Type::Update(v) => Some(v),
+        }
+    }
+}
+
+impl<K: Array, V: Array> Codec for Operation<K, V> {
+    fn write(&self, writer: &mut impl Writer) {
+        writer.write_fixed(&self.data);
+    }
+
+    fn read(reader: &mut impl Reader) -> Result<Self, CodecError> {
+        let value = reader.read_n_bytes(Self::LEN_ENCODED)?;
+        Self::try_from(value.as_ref()).map_err(|e| CodecError::Wrapped("Operation", e.into()))
+    }
+
+    fn len_encoded(&self) -> usize {
+        Self::LEN_ENCODED
+    }
+}
+
+impl<K: Array, V: Array> SizedCodec for Operation<K, V> {
+    const LEN_ENCODED: usize = K::LEN_ENCODED + 1 + V::LEN_ENCODED;
+}
+
+impl<K: Array, V: Array> Array for Operation<K, V> {
+    type Error = Error<K, V>;
+}
+
+impl<K: Array, V: Array> TryFrom<&[u8]> for Operation<K, V> {
+    type Error = Error<K, V>;
+
+    fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
+        if value.len() != Self::LEN_ENCODED {
+            return Err(Error::InvalidLength);
+        }
+
+        let _ = K::try_from(&value[..K::LEN_ENCODED]).map_err(|e| Error::InvalidKey(e))?;
+
+        match value[K::LEN_ENCODED] {
+            Self::UPDATE_CONTEXT => {
+                let _ = V::try_from(&value[K::LEN_ENCODED + 1..])
+                    .map_err(|e| Error::InvalidValue(e))?;
+            }
+            Self::DELETE_CONTEXT => {
+                // Check if the remaining bytes are all zeros
+                if !value[K::LEN_ENCODED + 1..].iter().all(|&b| b == 0) {
+                    return Err(Error::InvalidDeleteOp);
+                }
+            }
+            _ => {
+                return Err(Error::InvalidContextByte);
+            }
+        }
+
+        Ok(Self {
+            data: value.to_vec(),
+            _phantom: std::marker::PhantomData,
+        })
+    }
+}
+
+impl<K: Array, V: Array> TryFrom<&Vec<u8>> for Operation<K, V> {
+    type Error = Error<K, V>;
+
+    fn try_from(value: &Vec<u8>) -> Result<Self, Self::Error> {
+        Self::try_from(value.as_slice())
+    }
+}
+
+impl<K: Array, V: Array> TryFrom<Vec<u8>> for Operation<K, V> {
+    type Error = Error<K, V>;
+
+    fn try_from(value: Vec<u8>) -> Result<Self, Self::Error> {
+        Self::try_from(value.as_slice())
+    }
+}
+
+impl<K: Array, V: Array> AsRef<[u8]> for Operation<K, V> {
+    fn as_ref(&self) -> &[u8] {
+        &self.data
+    }
+}
+
+impl<K: Array, V: Array> Deref for Operation<K, V> {
+    type Target = [u8];
+    fn deref(&self) -> &[u8] {
+        &self.data
+    }
+}
+
+impl<K: Array, V: Array> Display for Operation<K, V> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self.to_type() {
+            Type::Deleted => write!(f, "[Key:{} <deleted>]", self.to_key()),
```

Comment:
nit: I've used lowercase for "keys" whenever I do this in other places. So, `[key:{} value:{}]`

Reply:
changed